### PR TITLE
Bug fixes in warm rain diagnostics (MODIS/CloudSat joint products)

### PR DESCRIPTION
**Pull request by takmichibata**
I have fixed some bugs in warm rain diagnostics (MODIS/CloudSat joint products) as reported in issue #39 and #41 from @dustinswales and @roj222.
Two files, `src/cosp.F90` and `src/cosp_stats.F90`, were edited due to the bug-fix.
I also attach the output data from test run after modifying the bugs, as raised in issue #42. The test was conducted by intel fortran compiler, and completed without any errors. For reference, I put the output data in `driver/data/outputs/UKMO/cosp2_output_um.ifort.bugfix.nc`.

Many thanks,
Tak

### CODE REVIEW

**Comment by RobertPincus** [comment]
@takmichibata Thanks for working on this. Before looking in detail at the code can make sure the automatic tests are passing? For example you can see large differences in https://github.com/CFMIP/COSPv2.0/pull/47/checks?check_run_id=712654828. @alejandrobodas can help with the automated tests.

**Comment by takmichibata** [comment]
@RobertPincus
Thank you for your check. The automatic tests seem to be failed at "Compare test output against known good output (KGO)". The detailed error messages suggest that differences in some variables from KGO are larger than the tolerances. The error variables are all from the warm rain diagnostics (i.e., npdf* and ncfodd*) because the bug fixes in this time affect these output diagnostics. The other variables seem to pass the automatic tests successfully against the KGO without errors.
If the modified codes are appropriate, could it be able to replace (update) the reference data, or check manually?

**Comment by RobertPincus** [comment]
@takmichibata Yes, we will want to replace the reference file as part of the pull request if the answers have changed because bugs have been fixed. Can you attach to the pull request some plots or similar showing that the results are now more correct?  @alejandrobodas feel free to weigh in on the process here.

**Comment by takmichibata** [comment]
@RobertPincus
Here I attach the plots showing CloudSat/CALIPSO total cloud fraction (black solid, left axis) and detected single-layered warm clouds (SLWC; color lines, right axis) for (a) initial version and (b) bug-fixed version.
In the initial version (a), the figure suggests that sampling of SLWCs does not work appropriately due to the inconsistency of vertical level between _frac_out_ and _dbze94_, when users specify `USE_VGRID=.true.`.
On the other hand, the revised version (b) detects more SLWCs, which seems consistent with CloudSat/CALIPSO cloud fraction and (c) MODIS low cloud cover in particular. In the most right point (loc=153 in X-axis), precipitating clouds are detected in the bug-fixed version as shown in the blue line, it is also consistent with "ptcloudsatflag3" which indicates "rain certain" (not shown). 
So I feel that the revised codes are now working appropriately.
Pull-request_n47-1

**Comment by RobertPincus** [comment]
@takmichibata Thanks for providing this figure. Since you are confident the answers are now correct, I suggest that you replace the "known-good-output" file in the repository so the automated tests pass. 

The COSP PMC will review the figures (we are even meeting in about 12 hours) and either ask for more evidence or simply accept the pull request. 

Thanks for being patient as we figure out new ways of working.

**Comment by alejandrobodas** [comment]
@takmichibata thanks for the making these changes. As @RobertPincus said, if you are confident about the new results, you can replace the file driver/data/outputs/UKMO/cosp2_output_um.gfortran.kgo.nc with the new outputs. You can retrieve the new outputs from the 'Artifacts' icon in the top-right area of: https://github.com/CFMIP/COSPv2.0/pull/47/checks?check_run_id=712654828

Once you push the changes to the repository, the tests should pass.

**Comment by takmichibata** [comment]
@alejandrobodas Thank you for telling me about the next step.
 Just to check, my output file in this revision is named as "`cosp2_output_um.ifort.bugfix.nc`" currently, is it no problem to upload? Or should I rename the new output as filename "`cosp2_output_um.gfortran.kgo.nc`" before uploading?

**Comment by alejandrobodas** [comment]
Hi @takmichibata yes, you will need to replace the current cosp2_output_um.gfortran.kgo.nc with the output of your test and push the changes.

**Comment by alejandrobodas** [comment]
Sorry @takmichibata I didn't explain the process very clearly. The file that you need to use is the output generated by the tests that are run in the github servers. You can download this file by clicking on 'Artifacts' in:
https://github.com/CFMIP/COSPv2.0/pull/47/checks?check_run_id=720558490

The image below shows where that link is located:
artifacts

Then, you can replace cosp2_output_um.gfortran.kgo.nc with the file you've downloaded, commit  and push the changes.

**Comment by takmichibata** [comment]
@alejandrobodas @RobertPincus Thank you very much for your kind supports. I replaced the existing KGO file with the new output I tested. The automatic test now goes well.

File: src/cosp_stats.F90 [hunk 1]
```diff
@@ -273,10 +273,10 @@ END SUBROUTINE COSP_LIDAR_ONLY_CLOUD
   !               2018.05.17 (T.Michibata): update for COSP2
   !               2018.09.19 (T.Michibata): modified I/O
   !               2018.11.22 (T.Michibata): minor revisions
-  ! * References: Suzuki et al. (JAS'10, JAS'11, JGR'13, GRL'13, JAS'15)
+  !               2020.05.27 (T.Michibata and X.Jing): bug-fix for frac_out dimsize
+  ! * References: Michibata et al. (GMD'19, GRL'20)
```
Comment (RobertPincus) [inline src/cosp_stats.F90:277]: Are you able to provide DOIs for these papers?

File: src/cosp.F90 [hunk 2]
```diff
@@ -351,9 +351,10 @@ function COSP_SIMULATOR(cospIN,cospgridIN,cospOUT,start_idx,stop_idx,debug)
          Lmodis_column,        & ! On/Off switch for column MODIS simulator
          Lrttov_column,        & ! On/Off switch for column RTTOV simulator (not used)
          Lradar_lidar_tcc,     & ! On/Off switch from joint Calipso/Cloudsat product
-         Lcloudsat_tcc,       & !
-         Lcloudsat_tcc2,      & !         
-         Llidar_only_freq_cloud  ! On/Off switch from joint Calipso/Cloudsat product
+         Lcloudsat_tcc,        & !
```
Comment (RobertPincus) [inline src/cosp.F90:354]: Please explain _tcc and _tcc2, and ideally pick a more informative name for _tcc2.

File: src/cosp.F90 [hunk 3]
```diff
@@ -384,7 +385,7 @@ function COSP_SIMULATOR(cospIN,cospgridIN,cospOUT,start_idx,stop_idx,debug)
     real(wp),dimension(:,:,:),allocatable :: &
        betamol_in,betamoli,pnormi,ze_toti,ze_noni
     real(wp),dimension(:,:,:),allocatable :: &
-         t_in,tmpFlip,ze_totFlip
+         t_in,tempI,frac_outI
```
Comment (RobertPincus) [inline src/cosp.F90:388]: Can these variable names be made more informative? Can you add comments explaining what they are?

File: src/cosp.F90 [hunk 4]
```diff
@@ -604,6 +606,7 @@ function COSP_SIMULATOR(cospIN,cospgridIN,cospOUT,start_idx,stop_idx,debug)
        Lmodis_subcolumn    = .true.
        Lcloudsat_column    = .true.
        Lcloudsat_subcolumn = .true.
+       Lcloudsat_modis_wr  = .true.
```
Comment (RobertPincus) [inline src/cosp.F90:609]: Please add a comment here explaining that _wr means "warm rain".

File: src/cosp_stats.F90 [hunk 5]
```diff
@@ -331,7 +331,10 @@ SUBROUTINE COSP_DIAG_WARMRAIN( Npoints, Ncolumns, Nlevels,           & !! in
     real(wp) :: cbtmh     !! diagnosed in-cloud optical depth
     real(wp), dimension(Npoints,Ncolumns,Nlevels) :: icod  !! in-cloud optical depth (ICOD)
     logical  :: octop, ocbtm, oslwc
-
+    integer, dimension(Npoints,Ncolumns,Nlevels) :: fracout_int  !! fracout (decimal to integer)
+    where (fracout .le. 0.5)                        fracout_int = SGCLD_CLR ! 0 = clear-sky
```
Comment (RobertPincus) [inline src/cosp_stats.F90:335]: Is there some reason not to use the `NINT()` intrinsic here? Also, perhaps you can explain in comments what's happening here.

File: src/cosp_stats.F90 (hunk 1 above)
Comment (takmichibata) [reply src/cosp_stats.F90:277]: Thanks for this suggestion. I added DOI information for the (selected) references.

File: src/cosp.F90 (hunk 2 above)
Comment (takmichibata) [reply src/cosp.F90:354]: These parameters are not from the CloudSat/MODIS warm rain diagnostics but from CloudSat/CALIPSO products, so I leave the variable name as it is. (I just only added indents here for legibility.)

File: src/cosp.F90 (hunk 3 above)
Comment (takmichibata) [reply src/cosp.F90:388]: These variable names are following to the CloudSat/CALIPSO joint-product. I added a comment line explaining these variables.

File: src/cosp.F90 (hunk 4 above)
Comment (takmichibata) [reply src/cosp.F90:609]: Yes, I added a comment line.

File: src/cosp_stats.F90 (hunk 5 above)
Comment (takmichibata) [reply src/cosp_stats.F90:335]: Thanks for this suggestion. I modified here using NINT() intrinsic function. I confirmed that the output result was not changed.

**Comment by takmichibata** [comment]
@RobertPincus Thanks for your code review in this revision. Please see my replies to your individual comments.

**Comment by alejandrobodas** [comment]
@takmichibata I've looked at the code and it looks good to me. I have one question regarding speed. Would you be able to quantify how much time is added to the calculations when these diagnostics are included? You can compare the outputs of "Time to run COSP" when you run the test in your local system with and without the warm rain diagnostics. In order to reduce the impact of variability, it is worth repeating the test a few times and report the average and fastest times. Thanks!

**Comment by takmichibata** [comment]
@alejandrobodas Thanks for your code review and suggestion. After this revision, I evaluated the computational cost regarding to the warm rain diagnostics based on two configurations; the one is that the all flags in `cosp2_output_nl.txt` are _.true._, and the other configuration is that the flags for warm rain diagnostics are _.false._. Both configurations were run for ten times, and their mean values are below:
**(a) with warm rain diagnostics**
Time to read in data: 8.9991003E-03
Time to initialize: 3.7991995E-03
Time to construct types: 1.3897900E-02
Time to compute optics: 7.8888200E-02
Time to run COSP: 6.2590301E-02
Total time: 0.1681747
Time to write to output: 1.6197799E-02

**(b) without warm rain diagnostics**
Time to read in data: 8.7989002E-03
Time to initialize: 3.8992996E-03
Time to construct types: 1.3598200E-02
Time to compute optics: 7.9188101E-02
Time to run COSP: 4.4093104E-02
Total time: 0.1495776
Time to write to output: 1.3298001E-02

The fastest values (Total time) are 0.167974 and 0.148978, respectively, suggesting that the warm rain diagnostics increase computational cost by approximately 12% under the ifort compiler.

**Pull request merged**